### DISCUSSION TITLE
Unable to control uav in gazebo

### DISCUSSION THREAD

**Discussion by afaroo01** [larics/icuas22_competition#44] [Q&A] [answered]
Hi, I am trying to deploy a navigation algorithm. But I'm getting the below-attached error and unable to control the uav. Any idea what's causing it and how to resolve this?

![IMG_20220407_103928](https://user-images.githubusercontent.com/64008660/162156869-5f1d1406-ef08-4457-aac8-3d9f003930c9.jpg)

**Comment by afaroo01** [comment]
@fpetric maybe you can tell?

**Reply by fpetric** [reply]
I think this is more @lmark1 territory. From what I can see, the issue is that your system does not know where Ardupilot's `sim_vehicle.py` script is located, which depending on the way you installed everything may be an issue with exporting PATH variable, or completely failed ardupilot installation. 
Maybe you can try adding the location of the `sim_vehicle.py` script to your PATH variable, it should be located in `<your_catkin_ws>/src/uav_ros_simulation/.gitman/ardupilot/Tools/autotest`

**Comment by lmark1** [comment, marked as answer]
Hi @afaroo01,

The challenge session runs with the [--no-rebuuild](https://github.com/larics/icuas22_competition/blob/86799e1b00bd91ff02ac0f50b603586c28bfc222/startup/challenge/session.yml#L43) argument, it therefore assumes the Copter firmware is built. 

The firmware comes prebuilt in the Docker images, so if you are using Docker please follow the troubleshooting instruction - [Update Docker Images](https://github.com/larics/icuas22_competition#update-docker-images) and rebuild your challenge image aswell.

If you are however building the system natively, you might need to build the Copter firmware manually by running:
```bash
roslaunch ardupilot_gazebo sim_vehicle.launch
```
or simply by removing the ```--no-rebuild``` flag.

Hope this helps!

**Reply by fpetric** [reply]
@lmark1 maybe we should update instructions somewhere for the non-docker people?

**Reply by lmark1** [reply]
Copter firmware build was recently added to the installation process, see https://github.com/larics/uav_ros_simulation/pull/20.

The troubleshooting guide is updated accordingly for the people with native installations prompting them to manually [re-run the installation process ](https://github.com/larics/icuas22_competition/blob/main/README.md#updating-native-installation).

**Reply by fpetric** [reply]
Thanks

**Reply by afaroo01** [reply]
Thanks @fpetric and @lmark1 , but even after resolving this issue now. The UAV is not being armed or able to send twist commands to control it. Don't know what is the reason? Will you provide a detailed instructions for manual setup for icuas 22 competition as well?

**Reply by fpetric** [reply]
The issue with arming may be due to Gazebo not loading the `ardupilot_gazebo` plugin, so then the ardupilot firmware does not receive the data it needs. Can you confirm you see this output in Gazebo launch window:
```
[Dbg] [ArduPilotPlugin.cc:488] ArduPilotPlugin::Load - Loading ArduPilotPlugin.

[Dbg] [ArduPilotPlugin.cc:90] Available sensor names:

[Dbg] [ArduPilotPlugin.cc:96] default::red::red/base_link::camera

[Dbg] [ArduPilotPlugin.cc:96] default::red::imu_link::imu_sensor

[Dbg] [dipole_magnet.cc:58] Loading DipoleMagnet plugin
```

With respect to twist messages, the UAV is not controled via `cmd_vel` topic but rather by publishing position and trajectory references, see [README.MD for more details](https://github.com/larics/icuas22_competition/blob/main/README.md#controlling-the-uav).
